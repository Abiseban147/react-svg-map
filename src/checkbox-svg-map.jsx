import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class CheckboxSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocations: []
		};

		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.isLocationSelected = this.isLocationSelected.bind(this);
	}

	handleLocationClick(event) {
		const location = event.target.attributes.name.value;

		this.setState(prevState => {
			// Copy old state
			let selectedLocations = [...prevState.selectedLocations];

			// Add or delete location
			if (event.target.attributes['aria-checked'].value === 'true') {
				const index = selectedLocations.indexOf(location);
				selectedLocations.splice(index, 1);
			} else {
				selectedLocations.push(location);
			}

			// Call onChange event handler
			this.props.onChange(selectedLocations);

			// Return new state
			return { selectedLocations };
		});

	}

	isLocationSelected(location) {
		return this.state.selectedLocations.indexOf(location) > -1;
	}

	render() {
		return (
			<SVGMap
				map={this.props.map}
				type="checkbox"
				onLocationClick={this.handleLocationClick}
				isLocationSelected={this.isLocationSelected}
				onLocationMouseOver={this.props.onLocationMouseOver}
				onLocationMouseOut={this.props.onLocationMouseOut}
				onLocationFocus={this.props.onLocationFocus}
				onLocationBlur={this.props.onLocationBlur}
				onChange={this.props.onChange} />
		);
	}
}

CheckboxSVGMap.propTypes = {
	onChange: PropTypes.func,

	// SVG map props
	map: PropTypes.node.isRequired,
	className: PropTypes.string,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
};

export default CheckboxSVGMap;
