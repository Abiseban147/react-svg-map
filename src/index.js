import SVGMap from './svg-map';
import Taiwan from './maps/taiwan';

export {
	SVGMap,
	Taiwan
};
