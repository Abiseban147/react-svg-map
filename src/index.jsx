import React from 'react';
import PropTypes from 'prop-types';

function SVGMap(props) {
	return (
		<svg className="svg-map" xmlns="http://www.w3.org/2000/svg" viewBox={props.map.viewBox} role="group" aria-label={props.map.label}>
			{
				props.map.locations.map(location => {
					return (<path
						id={location.id}
						className="svg-map__location"
						name={location.name}
						d={location.path}
						onMouseOver={props.onLocationMouseOver}
						onMouseOut={props.onLocationMouseOut}
						onClick={props.onLocationClick}
						onFocus={props.onLocationFocus}
						onBlur={props.onLocationBlur}
						tabIndex={props.tabIndex}
						role={props.type}
						aria-label={location.name}
						aria-checked={props.isLocationSelected && props.isLocationSelected(location)}
						key={location.id} />);
				})
			}
		</svg>
	);
}

SVGMap.propTypes = {
	map: PropTypes.shape({
		viewBox: PropTypes.string.isRequired,
		locations: PropTypes.arrayOf(
			PropTypes.shape({
				path: PropTypes.string.isRequired,
				name: PropTypes.string,
				id: PropTypes.string
			})
		).isRequired,
		label: PropTypes.string
	}).isRequired,
	tabIndex: PropTypes.string,
	type: PropTypes.string,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationClick: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	isLocationSelected: PropTypes.func
};

SVGMap.defaultProps = {
	tabIndex: '0', // Focusable locations
	type: 'none'
};

export default SVGMap;
