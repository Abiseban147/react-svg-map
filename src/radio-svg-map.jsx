import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class RadioSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocation: null
		};

		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.isLocationSelected = this.isLocationSelected.bind(this);
	}

	handleLocationClick(event) {
		const selectedLocation = event.target.attributes.name.value;

		// Change selected location
		this.setState({ selectedLocation });

		// Call onChange event handler
		this.props.onChange(selectedLocation);
	}

	isLocationSelected(location) {
		return this.state.selectedLocation === location;
	}

	render() {
		return (
			<SVGMap
				map={this.props.map}
				type="radio"
				onLocationClick={this.handleLocationClick}
				isLocationSelected={this.isLocationSelected}
				onLocationMouseOver={this.props.onLocationMouseOver}
				onLocationMouseOut={this.props.onLocationMouseOut}
				onLocationFocus={this.props.onLocationFocus}
				onLocationBlur={this.props.onLocationBlur}
				onChange={this.props.onChange} />
		);
	}
}

RadioSVGMap.propTypes = {
	onChange: PropTypes.func,

	// SVG map props
	map: PropTypes.node.isRequired,
	className: PropTypes.string,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
};

export default RadioSVGMap;
