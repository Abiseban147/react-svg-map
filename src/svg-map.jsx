import React from 'react';
import PropTypes from 'prop-types';

class SVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.svg = null;
		this.paths = null;

		// Select <svg> and <path> nodes
		this.setNodes = element => {
			this.svg = element.querySelector('svg');
			this.paths = element.querySelectorAll('path');
		};
	}

	componentDidMount() {
		// Add CSS class
		this.svg.setAttribute('class', this.props.className);

		for (let i = 0; i < this.paths.length; i++) {
			const path = this.paths[i];
			const name = path.getAttribute('name');

			// Add event listeners
			path.addEventListener('mouseover', this.props.onLocationMouseOver);
			path.addEventListener('mouseout', this.props.onLocationMouseOut);
			path.addEventListener('mousemove', this.props.onLocationMouseMove);
			path.addEventListener('click', this.props.onLocationClick);
			path.addEventListener('keydown', this.props.onLocationKeyDown);
			path.addEventListener('focus', this.props.onLocationFocus);
			path.addEventListener('blur', this.props.onLocationBlur);

			// Add attributes
			path.setAttribute('tabIndex', this.props.tabIndex);
			path.setAttribute('role', this.props.type);
			path.setAttribute('tabIndex', this.props.tabIndex);
			path.setAttribute('aria-label', name);

			// Add aria-checked attribute when needed
			if (this.props.isLocationSelected) {
				path.setAttribute('aria-checked', this.props.isLocationSelected(name));
			}
		}
	}

	componentDidUpdate() {
		// Update aria-checked when needed
		if (this.props.isLocationSelected) {
			for (let i = 0; i < this.paths.length; i++) {
				const path = this.paths[i];
				const name = path.getAttribute('name');

				path.setAttribute('aria-checked', this.props.isLocationSelected(name));
			}
		}
	}

	componentWillUnmount() {
		for (let i = 0; i < this.paths.length; i++) {
			const path = this.paths[i];

			// Remove event listeners
			path.removeEventListener('mouseover');
			path.removeEventListener('mouseout');
			path.removeEventListener('mousemove');
			path.removeEventListener('click');
			path.removeEventListener('keydown');
			path.removeEventListener('focus');
			path.removeEventListener('blur');
		}
	}

	render() {
		return (
			<div
				className="svg-map-wrapper"
				dangerouslySetInnerHTML={{ __html: this.props.map }}
				ref={this.setNodes}
			/>
		);
	}
}

SVGMap.propTypes = {
	map: PropTypes.node.isRequired,
	className: PropTypes.string,
	tabIndex: PropTypes.string,
	type: PropTypes.string,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationClick: PropTypes.func,
	onLocationKeyDown: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	isLocationSelected: PropTypes.func
};

SVGMap.defaultProps = {
	className: 'svg-map',
	tabIndex: '0', // Focusable locations
	type: 'none' // No role
};

export default SVGMap;
