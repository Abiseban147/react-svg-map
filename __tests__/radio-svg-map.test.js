import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { RadioSVGMap, Australia } from '../src';

jest.mock('react-dom', () => ({
	findDOMNode: () => ({
		querySelectorAll: jest.fn(),
	})
}));

// TODO: Create utility functions to avoid code duplication
describe('RadioSVGMap component', () => {
	const locationSelector = '#nsw';
	const previousLocationSelector = '#act';
	const nextLocationSelector = '#nt-groote-eylandt';
	const handleOnChange = jest.fn();
	let wrapper = null;
	let location = null;
	let previousLocation = null;
	let nextLocation = null;

	beforeEach(() => {
		// TODO: Use fake map to simplify tests?
		wrapper = mount(<RadioSVGMap map={Australia} onChange={handleOnChange} />);
		location = wrapper.find(locationSelector);
		previousLocation = wrapper.find(previousLocationSelector);
		nextLocation = wrapper.find(nextLocationSelector);
	});

	afterEach(() => {
		wrapper.unmount();
	});

	describe('Mouse navigation', () => {
		test('selects clicked location when not selected yet', () => {
			expect(location.props()['aria-checked']).toBeFalsy();

			location.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();
		});

		test('does not deselect clicked location when already selected', () => {
			location.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();

			location.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();
		});

		test('selects new clicked location and deselects former selected location', () => {
			location.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();
			expect(previousLocation.props()['aria-checked']).toBeFalsy();

			previousLocation.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);
			previousLocation = wrapper.find(previousLocationSelector);

			expect(location.props()['aria-checked']).toBeFalsy();
			expect(previousLocation.props()['aria-checked']).toBeTruthy();
		});

		test('makes location focusable when selected', () => {
			expect(location.props()['tabIndex']).toEqual('-1');

			location.simulate('click');
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['tabIndex']).toEqual('0');
		});
	});

	describe('Keyboard navigation', () => {
		test('selects focused unselected location when hitting spacebar', () => {
			expect(location.props()['aria-checked']).toBeFalsy();

			location.simulate('focus');
			location.simulate('keydown', { keyCode: 32 });
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();
		});

		test('does not deselect focused selected location when hitting spacebar', () => {
			location.simulate('focus');
			location.simulate('keydown', { keyCode: 32 });
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();

			location.simulate('focus');
			location.simulate('keydown', { keyCode: 32 });
			wrapper.update();
			location = wrapper.find(locationSelector);

			expect(location.props()['aria-checked']).toBeTruthy();
		});

		test('selects next/first location when hitting down/right arrow', () => {
			expect(location.props()['aria-checked']).toBeFalsy();
			expect(nextLocation.props()['aria-checked']).toBeFalsy();

			location.simulate('focus');
			location.simulate('keydown', { keyCode: 39 });
			wrapper.update();
			location = wrapper.find(locationSelector);
			nextLocation = wrapper.find(nextLocationSelector);

			expect(location.props()['aria-checked']).toBeFalsy();
			expect(nextLocation.props()['aria-checked']).toBeTruthy();
		});

		test('selects previous/last location when hitting up/left arrow', () => {
			expect(location.props()['aria-checked']).toBeFalsy();
			expect(previousLocation.props()['aria-checked']).toBeFalsy();

			location.simulate('focus');
			location.simulate('keydown', { keyCode: 37 });
			wrapper.update();
			location = wrapper.find(locationSelector);
			previousLocation = wrapper.find(previousLocationSelector);

			expect(location.props()['aria-checked']).toBeFalsy();
			expect(previousLocation.props()['aria-checked']).toBeTruthy();
		});
	});

	describe('Communication', () => {
		test('calls onChange handler when selecting location', () => {
			location.simulate('click');

			expect(handleOnChange).toHaveBeenCalledWith(location.getDOMNode());
		});
	});

	describe('Properties', () => {
		const map = {
			label: 'label',
			viewBox: 'viewBox',
			locations: [{
				name: 'name',
				id: 'id',
				path: 'path'
			}]
		};

		test('displays map with default props', () => {
			const component = renderer.create(<RadioSVGMap map={map} />);
			const tree = component.toJSON();

			expect(tree).toMatchSnapshot();
		});

		test('displays map with custom props', () => {
			const eventHandler = () => 'eventHandler';
			const component = renderer.create(
				<RadioSVGMap map={map}
					className="className"
					locationClassName="locationClassName"
					onLocationMouseOver={eventHandler}
					onLocationMouseOut={eventHandler}
					onLocationMouseMove={eventHandler}
					onLocationFocus={eventHandler}
					onLocationBlur={eventHandler}
					onChange={eventHandler}
				/>
			);
			const tree = component.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});
});