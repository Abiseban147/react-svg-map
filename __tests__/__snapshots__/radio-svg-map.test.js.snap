// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioSVGMap component Properties displays map with custom props 1`] = `
<svg
  aria-label="label"
  className="className"
  role="radiogroup"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    aria-checked={null}
    aria-label="name"
    className="locationClassName"
    d="path"
    id="id"
    name="name"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="radio"
    tabIndex="0"
  />
</svg>
`;

exports[`RadioSVGMap component Properties displays map with default props 1`] = `
<svg
  aria-label="label"
  className="svg-map"
  role="radiogroup"
  viewBox="viewBox"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    aria-checked={null}
    aria-label="name"
    className="svg-map__location"
    d="path"
    id="id"
    name="name"
    onBlur={undefined}
    onClick={[Function]}
    onFocus={undefined}
    onKeyDown={[Function]}
    onMouseMove={undefined}
    onMouseOut={undefined}
    onMouseOver={undefined}
    role="radio"
    tabIndex="0"
  />
</svg>
`;
